source/extensions/common/async_files/async_file_manager_thread_pool.cc

```diff
@@ -0,0 +1,260 @@
+  }
+
+private:
+  const std::string path_;
+};
+
+class ActionOpenExistingFile : public ActionWithFileResult {
+public:
+  ActionOpenExistingFile(AsyncFileManagerThreadPool& manager, absl::string_view filename,
+                         AsyncFileManager::Mode mode,
+                         std::function<void(absl::StatusOr<AsyncFileHandle>)> on_complete)
+      : ActionWithFileResult(manager, on_complete), filename_(filename), mode_(mode) {}
+
+  absl::StatusOr<AsyncFileHandle> executeImpl() override {
+    int fd = posix().open(filename_.c_str(), openFlags());
+    if (fd == -1) {
+      return statusAfterFileError();
+    }
+    return std::make_shared<AsyncFileContextThreadPool>(manager_, fd);
+  }
+
+private:
+  int openFlags() const {
+    switch (mode_) {
+    case AsyncFileManager::Mode::ReadOnly:
+      return O_RDONLY;
+    case AsyncFileManager::Mode::WriteOnly:
+      return O_WRONLY;
+    case AsyncFileManager::Mode::ReadWrite:
+    default:
+      return O_RDWR;
+    }
+  }
+  const std::string filename_;
+  const AsyncFileManager::Mode mode_;
+};
+
+class ActionUnlink : public AsyncFileActionWithResult<absl::Status> {
+public:
+  ActionUnlink(const PosixFileOperations& posix, absl::string_view filename,
+               std::function<void(absl::Status)> on_complete)
+      : AsyncFileActionWithResult(on_complete), posix_(posix), filename_(filename) {}
+
+  absl::Status executeImpl() override {
+    int result = posix_.unlink(filename_.c_str());
+    if (result == -1) {
+      return statusAfterFileError();
+    }
+    return absl::OkStatus();
+  }
+
+private:
+  const PosixFileOperations& posix_;
+  const std::string filename_;
+};
+
+} // namespace
+
+std::function<void()> AsyncFileManagerThreadPool::createAnonymousFile(
+    absl::string_view path, std::function<void(absl::StatusOr<AsyncFileHandle>)> on_complete) {
+  return enqueue(std::make_shared<ActionCreateAnonymousFile>(*this, path, on_complete));
+}
+
+std::function<void()> AsyncFileManagerThreadPool::openExistingFile(
+    absl::string_view filename, Mode mode,
+    std::function<void(absl::StatusOr<AsyncFileHandle>)> on_complete) {
+  return enqueue(std::make_shared<ActionOpenExistingFile>(*this, filename, mode, on_complete));
+}
+
+std::function<void()>
+AsyncFileManagerThreadPool::unlink(absl::string_view filename,
+                                   std::function<void(absl::Status)> on_complete) {
+  return enqueue(std::make_shared<ActionUnlink>(posix(), filename, on_complete));
+}
+
+class AsyncFileManagerThreadPoolFactory : public AsyncFileManagerFactory {
+  bool shouldUseThisFactory(const AsyncFileManagerConfig& config) const override {
+    return config.thread_pool_size.has_value();
+  }
+  std::unique_ptr<AsyncFileManager> create(const AsyncFileManagerConfig& config) const override {
+    return std::make_unique<AsyncFileManagerThreadPool>(config);
+  }
+};
+static AsyncFileManagerThreadPoolFactory registered;
```

Comment:
Wouldn't it make sense to register and to configure this factory as an Envoy bootstrap extension?

Reply:
So the original idea here was this library could be usable even without being tied into envoy stuff, but given that it has now reached the point of being fairly thoroughly tied in, it probably does make more sense to instantiate things through an extension. However, it *also* needs a way for the clients to use it, which was originally part of the next PR (registering a singleton factory). I guess at this point it makes more sense to smash those two parts together, and roll a singleton factory into this PR.

Part of the intent of *this* design was that you could have a build that just completely excludes some targets from builds that don't support them, and the absence of the self-registering presence would magically make that feature not supported. But I think maybe the Envoy way is more to build everything that can build, and just have the unsupported option explicitly panic or error. With OsSysCalls now supporting that (so that this can probably also build for Windows now), I can get on board with that paradigm.

So work in progress migrating this to OsSysCalls, a config from a proto (since that's what it's going to have to be later anyway), and a singleton factory.